dpnp/dpnp_iface_linearalgebra.py

```diff
@@ -269,63 +285,364 @@ def matmul(x1, x2, out=None, **kwargs):
+            x2_is_2D = dpnp.all(dpnp.array(x2_shape[:-2]) == 1)
+
+            # makes the dimension of input the same by adding new axis
+            if x1_ndim != x2_ndim:
+                diff = abs(x1_ndim - x2_ndim)
+                if x1_ndim < x2_ndim:
+                    x1 = x1.reshape((1,) * diff + x1.shape)
+                    x1_ndim = x1.ndim
+                    x1_shape = x1.shape
+                else:
+                    x2 = x2.reshape((1,) * diff + x2.shape)
+                    x2_ndim = x2.ndim
+                    x2_shape = x2.shape
+
+            # examining the option to align inputs when their shapes differ but they are 1-D in some dimensions.
+            tmp_shape = list(x1_shape[:-2])
+            for i in range(x1_ndim - 2):
+                if x1_shape[i] != x2_shape[i]:
+                    if x1_shape[i] == 1:
+                        tmp_shape[i] = x2_shape[i]
+                        # If the `x1` array is inherently 2D, there's no need to duplicate the data for the 1-D dimension; GEMM handles it automatically.
+                        if not x1_is_2D:
+                            x1 = dpnp.repeat(x1, x2_shape[i], axis=i)
+                    elif x2_shape[i] == 1:
+                        tmp_shape[i] = x1_shape[i]
+                        if not x2_is_2D:
+                            x2 = dpnp.repeat(x2, x1_shape[i], axis=i)
+                    else:
+                        raise ValueError(
+                            "operands could not be broadcast together with remapped shapes."
+                        )
+            x1_shape = x1.shape
+            x2_shape = x2.shape
+            res_shape = tuple(tmp_shape) + (x1_shape[-2], x2_shape[-1])
+
+        # calculate results
+        res_usm_type = du.get_coerced_usm_type([x1.usm_type, x2.usm_type])
+        result = dpnp.empty(
+            res_shape,
+            dtype=gemm_dtype,
+            usm_type=res_usm_type,
+            sycl_queue=exec_q,
+        )
+        if result.size == 0:
             pass
         else:
-            if 0:
-                """
-                Cost model checks
-                """
-
-                array1_size = x1_desc.size
-                array2_size = x2_desc.size
-                cost_size = 4096  # 2D array shape(64, 64)
-
-                if (x1_desc.dtype == dpnp.float64) or (
-                    x1_desc.dtype == dpnp.float32
-                ):
-                    """
-                    Floating point types are handled via original math library better than SYCL math library
-                    """
-                    cost_size = 262144  # 2D array shape(512, 512)
-
-                if (array1_size > cost_size) and (array2_size > cost_size):
-                    return dpnp_matmul(x1_desc, x2_desc, out)
+            if x1.size == 0 or x2.size == 0:
+                result = dpnp.zeros(
+                    res_shape, dtype=gemm_dtype, sycl_queue=exec_q
```

Comment:
```suggestion
                    res_shape, dtype=gemm_dtype, sycl_queue=exec_q, usm_type=res_usm_type
```

Additionally please add a test for this case in `test_usm_type.py`